Ilariae/clean ntt (#466)
wormhole-foundation/wormhole-docs: products/native-token-transfers/guides/deploy-to-evm.md

This guide walks you through deploying NTT on EVM chains, including setting up dependencies, configuring token compatibility, and using the NTT CLI to deploy in hub-and-spoke or burn-and-mint mode.

## Prerequisites

Before you begin, ensure you have the following:

- An existing ERC-20 token deployed to the [EVM chains](/docs/products/reference/supported-networks/#ntt){target=\_blank} you plan to use
- Make sure you have the NTT CLI installed and a project initialized. Follow the below steps or see the [Get Started guide](/docs/products/native-token-transfers/get-started/#install-ntt-cli){target=\_blank}:

    ???- interface "Install the NTT CLI and Scaffold a New Project"
    
        --8<-- 'text/products/native-token-transfers/guides/install-ntt-project.md'

> 

Reviewer: I'm noticing the Solana deployment guide has an "Overview of the Deployment Process" included and this one does not. I know we have Deployment Process steps in the NTT/Overview so I'm not saying they should also go here, just that this is one of the things that should be decided on and be consistent between the two deployment guides.

This guide walks you through deploying NTT on EVM chains, including setting up dependencies, configuring token compatibility, and using the NTT CLI to deploy in hub-and-spoke or burn-and-mint mode.

## Prerequisites

Before deploying NTT on EVM chains, ensure you have the following prerequisites:

- [Node.js and npm installed](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm){target=\_blank}
- [Bun installed](https://bun.sh/){target=\_blank}
- A wallet private key with tokens on supported chains
- ERC-20 tokens already deployed on the source and destination chains

## Overview of the Deployment Process

Deploying NTT on EVM chains follows a structured process:

1. **Choose your token setup**: Use an existing ERC-20 token or deploy a new one.

    ???- interface "Deploy an ERC-20 Token on EVM"
        Use the [example NTT token repository](https://github.com/wormhole-foundation/example-ntt-token){target=\_blank} to deploy a basic ERC-20 token contract on testnet.

>         1. **Install Foundry**: Install the [Forge CLI](https://getfoundry.sh/introduction/installation/){target=\_blank}

Reviewer 2: ```suggestion
        1. **Install Foundry**: Install the [Forge CLI](https://getfoundry.sh/introduction/installation/){target=\_blank}.
```

Please check all the list items (in this set of steps in particular to) make sure they end with a period.

>         1. **Install Foundry** - install the [Forge CLI](https://getfoundry.sh/introduction/installation/){target=\_blank}

Reviewer 2: ```suggestion
        1. **Install Foundry**: Install the [Forge CLI](https://getfoundry.sh/introduction/installation/){target=\_blank}.
```

Pls check rest of list items

        2. **Clone the repository**: Fetch the example contract repository

            ```bash
            git clone https://github.com/wormhole-foundation/example-ntt-token.git
            cd example-ntt-token
            ```
        
        3. **Deploy the token contract**: Deploy to testnet with your preferred name, symbol, minter, and owner addresses

            ```bash
            forge create --broadcast \
                --rpc-url INSERT_RPC_URL \
                --private-key INSERT_YOUR_PRIVATE_KEY \
                src/PeerToken.sol:PeerToken \
                --constructor-args "INSERT_TOKEN_NAME" "INSERT_TOKEN_SYMBOL" INSERT_MINTER_ADDRESS INSERT_OWNER_ADDRESS
            ```

        4. **Mint tokens**: Send tokens to your address

            ```bash
            cast send INSERT_TOKEN_ADDRESS \
                "mint(address,uint256)" \
                INSERT_RECIPIENT_ADDRESS \
                INSERT_AMOUNT_IN_WEI \
                --private-key INSERT_YOUR_PRIVATE_KEY \
                --rpc-url INSERT_RPC_URL
            ```

        !!! note
            This token uses 18 decimals by default. All minting values must be specified in `wei` (1 token = 10^18).

> 2. **Choose your deployment model**: Choose a deployment model. Wormhole’s NTT framework supports two [deployment models](/docs/products/native-token-transfers/overview#deployment-models){target=\_blank}: burn-and-mint and hub-and-spoke.

Reviewer 2: ```suggestion
2. **Choose your deployment model**: Choose a deployment model. The NTT framework supports two [deployment models](/docs/products/native-token-transfers/overview#deployment-models){target=\_blank}: burn-and-mint and hub-and-spoke.
```

    ??? interface "Burn-and-Mint"

        Tokens must implement the following non-standard ERC-20 functions:

        - `burn(uint256 amount)`
        - `mint(address account, uint256 amount)`

        These functions aren't part of the standard ERC-20 interface. Refer to the [`INttToken` interface](https://github.com/wormhole-foundation/native-token-transfers/blob/main/evm/src/interfaces/INttToken.sol){target=\_blank} for all required functions, errors, and events.

        ??? interface "`INttToken` Interface"
            ```solidity
            --8<-- 'code/products/native-token-transfers/guides/deploy-to-evm/INttToken.sol'
            ```

        You’ll also need to set mint authority to the relevant `NttManager` contract.

>     ??? interface "Hub-and-Spoke Mode"

Reviewer 2: ```suggestion
    ??? interface "Hub-and-Spoke"
```

To match the burn and mint title

        Tokens only need to be ERC-20 compliant. The hub chain serves as the source of truth for supply consistency, while only spoke chains need to support minting and burning. For example, if Ethereum is the hub and Polygon is a spoke:

        - Tokens are locked on Ethereum

>         - Tokens are minted or burned on Polygon

Reviewer 2: ```suggestion
        - Tokens are locked on Ethereum.
        - Tokens are minted or burned on Polygon.
```

            ntt init Testnet
            ```

## Deploy and Configure NTT

Once you've set up NTT, proceed with adding your EVM chains and deploying contracts.

> 1. **Environment Setup**: Ensure you have set up your environment correctly, open your terminal and run the `export` commands:

Reviewer: ```suggestion
1. **Environment Setup**: Ensure you have set up your environment correctly, open your terminal, and run the `export` commands:
```

    - Name of each chain.
    - Version of NTT to deploy (use `--latest` for the latest contract versions).
    - Mode - either `burning` or `locking`.
    - Your token contract address.

    While not recommended, you can pass the `-skip-verify` flag to the `ntt add-chain` command if you want to skip contract verification.

3. **Verify deployment status**: After deployment, check if your `deployment.json` file matches the on-chain configuration using the following command:

    ```bash
    ntt status
    ```

    If needed, sync your local configuration with the on-chain configuration:

    ```bash
    ntt pull
    ```

4. **Configure rate limits**: Set up inbound and outbound rate limits. By default, limits are set to 0 and must be updated before deployment. For EVM chains, values must be set using 18 decimals.

    Open your `deployment.json` file and adjust the values based on your use case:

      ```json
      "inbound": {
          "Arbitrum": "1000.000000000000000000" // inbound limit from Arbitrum to Ethereum
      }
      "outbound": {
          "Ethereum": "1000.000000000000000000" // outbound limit from Ethereum to Arbitrum
      }
      ```

>     This ensures your rate limits align with the token’s precision on each chain, preventing mismatches that could block or miscalculate transfers. Before setting these values, confirm your token’s decimals on each chain by checking the token contract on the relevant block explorer.

Reviewer: ```suggestion
    This configuration ensures your rate limits align with the token’s precision on each chain, preventing mismatches that could block or miscalculate transfers. Before setting these values, confirm your token’s decimals on each chain by checking the token contract on the relevant block explorer.
```

| `outboundLimit`         | Computed               | Auto-detected/Hardcoded             | Similar| Relative to rate limit             |

## Deploy and Configure NTT

Once you've set up NTT, proceed with adding your EVM chains and deploying contracts.

1. **Environment Setup**: Ensure you have set up your environment correctly:

    ```bash

>     export ETH_PRIVATE_KEY=INSERT_PRIVATE_KEY

Reviewer 2: Where is this supposed to go?

    ```

> 2. **Deploy NTT to EVM**: Add each chain you'll be deploying to. The following example demonstrates configuring NTT in burn-and-mint mode on Ethereum Sepolia and Arbitrum Sepolia:

Reviewer 2: ```suggestion
2. **Deploy NTT to EVM**: Add each chain you'll be deploying to using the `ntt add-chain` command. The following example demonstrates configuring NTT in burn-and-mint mode on Ethereum Sepolia and Arbitrum Sepolia:
```

    ```bash
    --8<-- 'code/products/native-token-transfers/guides/deploy-to-evm/initialize.txt'
    ```

>     The `ntt add-chain` command takes the following parameters:

Reviewer 2: Have you tried putting the command details in an interface admonition? Might look a little cleaner, might not.

Author: I put it in a terminal output and it looks worse to me but you check it and lmk what u think

Reviewer 2: Not a terminal, but an interface admonition. 

My comment initially was about the parameters you listed out

Reviewer 2: When creating stuff like this, feel free to try out different approaches to see what looks best

    - Name of each chain.
    - Version of NTT to deploy (use `--latest` for the latest contract versions).
    - Mode - either `burning` or `locking`.
    - Your token contract address.

    While not recommended, you can pass the `-skip-verify` flag to the `ntt add-chain` command if you want to skip contract verification.

3. **Verify deployment status** - after deployment, check if your `deployment.json` file matches the on-chain configuration using the following command:

    ```bash
    ntt status
    ```

    If needed, sync your local configuration with the on-chain configuration:

    ```bash
    ntt pull
    ```

> 4. **Configure rate limits**: Set up rate limits with the appropriate number of decimals, depending on the specific chain. For example:

Reviewer 2: I think this was modified in the solana page because it had details for the EVM config in the solana page and vice versa. We should only focus on EVM since they're deploying EVM in this guide and then the Solana guide will have the necessary details there

Reviewer 3: I think this part is fine. IMO it's good to have the two examples. Let's say you're integrating NTT to your ERC20 token and you want to set both inbound and outbound limits. Showing how to set it up for transfers coming from Solana could be helpful. 
Maybe I already forgot the reason why we changed it in the Solana page haha

Reviewer 2: We changed it because they're not configuring EVM when following the Solana guide. They'll configure the EVM stuff when they follow the EVM guide.

    ```

    The `ntt add-chain` command takes the following parameters:

    - Name of each chain.
    - Version of NTT to deploy (use `--latest` for the latest contract versions).
    - Mode - either `burning` or `locking`.
    - Your token contract address.

    While not recommended, you can pass the `-skip-verify` flag to the `ntt add-chain` command if you want to skip contract verification.

3. **Verify deployment status**: After deployment, check if your `deployment.json` file matches the on-chain configuration using the following command:

    ```bash
    ntt status
    ```

    If needed, sync your local configuration with the on-chain configuration:

    ```bash
    ntt pull
    ```

4. **Configure rate limits**: Set up rate limits with the appropriate number of decimals, depending on the specific chain. For example, if your token has 18 decimals on Ethereum and 6 decimals on Arbitrum::

    To Ethereum, the limits are set with 18 decimal places:
      ```json
      "inbound": {
          "Arbitrum": "1000.000000000000000000" // inbound limit from Arbitrum to Ethereum
      }
      ```

>     To Arbitrum, the limits are set with 6 decimal places:

Reviewer 2: So to me this reads as ultimately, you'll need to have something like this in a single file (since we're working on the configuration file related to the deployment right now):

```
      "inbound": {
          "Arbitrum": "1000.000000000000000000", // inbound limit from Arbitrum to Ethereum
          "Ethereum": "1000.000000" // inbound limit from Ethereum to Arbitrum
      }

```

Author: yeah u can put it together since they're both inbounds, I'll change one to outbound

Author: but @martin0995 please check this rate limits part with the decimals

Reviewer 3: https://sepolia.arbiscan.io/address/0xCeC6FB4F352bf3DC2b95E1c41831E4D2DBF9a35D#readProxyContract#F26

    export SEPOLIA_SCAN_API_KEY=INSERT_ETHERSCAN_SEPOLIA_API_KEY
    export ARBITRUMSEPOLIA_SCAN_API_KEY=INSERT_ARBISCAN_SEPOLIA_API_KEY
    ```

2. **Deploy NTT to EVM**: Add each chain you'll be deploying to using the `ntt add-chain` command. The following example demonstrates configuring NTT in burn-and-mint mode on Ethereum Sepolia and Arbitrum Sepolia:

    ```bash
    --8<-- 'code/products/native-token-transfers/guides/deploy-to-evm/initialize.txt:4'
    ```

    The `ntt add-chain` command takes the following parameters:

    - Name of each chain.
    - Version of NTT to deploy (use `--latest` for the latest contract versions).
    - Mode - either `burning` or `locking`.
    - Your token contract address.

    While not recommended, you can pass the `-skip-verify` flag to the `ntt add-chain` command if you want to skip contract verification.

3. **Verify deployment status**: After deployment, check if your `deployment.json` file matches the on-chain configuration using the following command:

    ```bash
    ntt status
    ```

    If needed, sync your local configuration with the on-chain configuration:

    ```bash
    ntt pull
    ```

> 4. **Configure rate limits**: Set up rate limits with the appropriate number of decimals, depending on the specific chain. For example, if your token has 18 decimals on Ethereum and 6 decimals on Arbitrum::

Reviewer 2: ```suggestion
4. **Configure rate limits**: Set up rate limits with the appropriate number of decimals, depending on the specific chain. For example, if your token has 18 decimals on Ethereum and 6 decimals on Arbitrum:
```

- If you followed the [`INttToken`](https://github.com/wormhole-foundation/native-token-transfers/blob/main/evm/src/interfaces/INttToken.sol){target=\_blank} interface, you can execute the `setMinter(address newMinter)` function
    ```json

>     cast send $TOKEN_ADDRESS "setMinter(address)" $NTT_MANAGER_ADDRESS --keystore /path/to/keystore.json --password <(echo "YOUR_PASSWORD") --rpc-url $YOUR_RPC_URL 

Reviewer: I can't get to it down below here but, that "!!! important" should be updated to be a Next Steps section instead

      ```

    This initial configuration ensures that the rate limits are correctly represented for each chain's token precision

> 5. **Push the final deployment**: Once rate limits are set, sync the on-chain configuration with local changes made to your `deployment.json` file

Reviewer 2: ```suggestion
5. **Push the final deployment**: Once rate limits are set, sync the on-chain configuration with local changes made to your `deployment.json` file.
```

By default, NTT transfers to EVM blockchains support automatic relaying via the Wormhole relayer, which doesn't require the user to perform a transaction on the destination chain to complete the transfer.

## NTT Manager Deployment Parameters

> This table compares the configuration parameters available when deploying the NTT Manager using the CLI versus a manual deployment with a Forge script. It highlights which options are configurable via each method, whether values are auto-detected or hardcoded, and includes additional comments to help guide deployment decisions.

Reviewer: ```suggestion
This table compares the configuration parameters available when deploying the NTT Manager using the CLI versus those available during a manual deployment with a Forge script. It highlights which options are configurable via each method, whether values are auto-detected or hardcoded, and includes additional comments to help guide deployment decisions.
```

By default, NTT transfers to EVM blockchains support automatic relaying via the Wormhole relayer, which doesn't require the user to perform a transaction on the destination chain to complete the transfer.

## Where to Go Next

<div class="grid cards" markdown>

-   :octicons-tools-16:{ .lg .middle } **Test Your Deployment**

    ---

    Follow the NTT Post Deployment Guide for integration examples and testing instructions.

    [:custom-arrow: Test Your NTT deployment](/docs/products/native-token-transfers/guides/post-deployment/){target=\_blank}

> -   :octicons-tools-16:{ .lg .middle } **Deploy NTT with Launchpad**

Reviewer 2: Not sure this one makes the most sense

By default, NTT transfers to EVM blockchains support automatic relaying via the Wormhole relayer, which doesn't require the user to perform a transaction on the destination chain to complete the transfer.

## Where to Go Next

> <div class="grid cards" markdown>

Reviewer 2: I think it's fine to have kept the Solana deployment card here, but it just didn't make sense to tell users to go to the NTT launchpad since since that's EVM chains and this is the EVM deployment guide, you know what I mean?
